Remote store implementation

anwar6953 opened the pull request:
I think its useful to review the commits separately:
1. Moves methods from Store interface into a RuntimeStore interface. These are methods that would be needed from program runtime (i.e. when a workflow or mapreduce is running, it needs to perform some updates to the meta store).
2. Create an implementation of RuntimeStore that makes HTTP requests to app fabric, so that the CDAP system can perform the actual store operations.
3. Update MapReduceProgramRunner, SparkProgramRunner, and WorkflowProgramRunner to use RemoteRuntimeStore.

We need this because CDAP programs (workflow/mapreduce/etc) may no longer have permission to write to system HBase tables, once impersonation is implemented.
See: https://wiki.cask.co/display/CE/Secure+Impersonation+-+Security+3.5#SecureImpersonation-Security3.5-ProblemsEncountered
https://issues.cask.co/browse/CDAP-6250

Pending (which will come in a separate PR):
1. Remote versions of usage and lineage stores.
2. Scalability of this service/handler.
3. Not using deprecated `Id` classes in new code. This will come as a separate change, since the new interface shares methods with existing interface and implementation.

Build: http://builds.cask.co/browse/CDAP-DUT4241

anwar6953 committed:
Separate RuntimeStore from Store interface.

anwar6953 committed:
Implement RemoteRuntimeStore. Exposes http endpoints for writing to system tables, as necessary, during program runtime.

anwar6953 committed:
Update MapReduceProgramRunner, SparkProgramRunner, and WorkflowProgramRunner to use RemoteRuntimeStore.

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/app/store/RuntimeStore.java:
```diff
@@ -0,0 +1,123 @@
+import javax.annotation.Nullable;
+
+/**
+ * Manages the runtime lifecycle of a {@link Program}.
```
Not really "manages" lifecycle, but a store for runtime information of programs.

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/app/store/RuntimeStore.java:
```diff
@@ -0,0 +1,123 @@
+   * @param pid       run id
+   * @param startTime start timestamp in seconds; if run id is time-based pass the time from the run id
+   */
+  @VisibleForTesting
```
Remove this. All methods in interface are public and is part of the interface contract.

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/app/store/RuntimeStore.java:
```diff
@@ -0,0 +1,123 @@
+   * @param id      id of the program
+   * @param pid     run id
+   */
+  void setResume(Id.Program id, String pid);
```
Use the new `ProgramId` for new class. As I mentioned before, you can keep the old method in the implementation.

anwar6953 replied:
I've already have changed it in a separate branch (wip here: https://github.com/caskdata/cdap/commit/ec07215b755cb52f41040686a13ce57b3105d5be), to use the new classes, but for simplicity, I will open that as a separate PR (it is pending item #3 in this PR description).

anwar6953 committed:
Address comments

bdmogal commented on cdap-app-fabric/src/main/java/co/cask/cdap/app/store/RuntimeStore.java:
```diff
@@ -0,0 +1,112 @@
+   * Compare and set operation that allow to compare and set expected and update status.
+   * Implementation of this method should guarantee that the operation is atomic or in transaction.
+   *
+   * @param id              Info about program
```
this may just be copied over, but let's fix the javadocs in this class to have a single space between parameter name and description.

Also, this is the program Id, not info about program?

Nishith replied:
Name doesn't match the definition, use something like programInfo?

anwar6953 replied:
@Nishith will change the description instead

@bdmogal So, we don't want this style?

![image](https://cloud.githubusercontent.com/assets/2440977/16347009/30ad822e-39ff-11e6-8e65-f0f5c8e8c8c1.png)

bdmogal replied:
I think we had decided against it. Maybe @awholegunch can confirm?

awholegunch replied:
Yes, our Javadoc style is to use a sing;e space between the parameter name and description. Otherwise, you change something, and a lot of time is spent reformatting.

That trailing period on `name of the OutputFormat class` should be deleted. But is that description correct? Should it be:

```
name of the outputFormat class
```

or should the parameter be:

```
outputFormatClassname
```

bdmogal commented on cdap-app-fabric/src/main/java/co/cask/cdap/app/store/RuntimeStore.java:
```diff
@@ -0,0 +1,112 @@
+   * @param expectedStatus  The expected value
+   * @param updateStatus    The new value
+   */
+  void compareAndSetStatus(Id.Program id, String pid, ProgramRunStatus expectedStatus, ProgramRunStatus updateStatus);
```
let's rename the second parameter to runId? Also, is there an argument against using Id.Run/ProgramRunId here instead of two separate parameters? Applies to all methods in this interface?

Nishith replied:
CAS methods generally use expected and new prefix to identify the two values.

anwar6953 replied:
@bdmogal will change that as part of separate PR (task #3 in pending items).

@Nishith updated.

bdmogal commented on cdap-app-fabric/src/main/java/co/cask/cdap/app/store/RuntimeStore.java:
```diff
@@ -0,0 +1,112 @@
+   * @param runtimeArgs the runtime arguments for this program run
+   * @param systemArgs the system arguments for this program run
+   */
+  void setStart(Id.Program id, String pid, long startTime, @Nullable String twillRunId,
```
ditto

bdmogal commented on cdap-app-fabric/src/main/java/co/cask/cdap/app/store/RuntimeStore.java:
```diff
@@ -0,0 +1,112 @@
+  /**
+   * Logs start of program run.
+   *
+   * @param id         Info about program
```
program id

bdmogal commented on cdap-app-fabric/src/main/java/co/cask/cdap/app/store/Store.java:
```diff
@@ -48,20 +48,9 @@
 /**
  * {@link Store} operates on a {@link Program}. It's responsible
- * for managing the non-runtime lifecycle of a {@link Program}
+ * for managing the lifecycle of a {@link Program}.
```
so now it only manages the non-runtime lifecycle right? this may increase ambiguity though, should we just change it to manages app/program metadata?

bdmogal commented on cdap-app-fabric/src/main/java/co/cask/cdap/app/store/Store.java:
```diff
@@ -48,20 +48,9 @@
-   * @param updateStatus    The new value
-   */
-  void compareAndSetStatus(Id.Program id, String pid, ProgramRunStatus expectedStatus, ProgramRunStatus updateStatus);
+public interface Store extends RuntimeStore {
```
why does Store have to extend RuntimeStore?

anwar6953 replied:
we also want the methods defined in RuntimeStore to be available in Store class

bdmogal commented on cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/meta/RemoteRuntimeStoreHandler.java:
```diff
@@ -0,0 +1,171 @@
+import javax.ws.rs.Path;
+
+/**
+ * The {@link co.cask.http.HttpHandler} for handling REST calls to meta stores.
```
nit: use import + simple name?

bdmogal replied:
also, ... calls to runtime store?

bdmogal commented on cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/meta/package-info.java:
```diff
@@ -15,7 +15,6 @@
 
 /**
- * Classes to handle data fabric requests.
+ * Classes to handle REST calls to meta stores.
```
what are meta stores?

anwar6953 replied:
changed to 'meta data stores'

bdmogal commented on cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/util/AbstractAppFabricHttpHandler.java:
```diff
@@ -112,20 +112,11 @@ protected int getInstances(HttpRequest request) throws BadRequestException {
   protected Map<String, String> decodeArguments(HttpRequest request) throws JsonSyntaxException {
-    ChannelBuffer content = request.getContent();
-    if (!content.readable()) {
+    Map<String, String> args = parseBody(request, STRING_MAP_TYPE);
```
nice cleanup. So the only thing that we lose because of this is the message? Worth it to have `parseBody` accept an error message too?

bdmogal replied:
Actually, nvm. that seems like just a log message, not an error message propagated to users. Where `parseBody` catches a runtime exception on line 107 above, should we throw throw a `BadRequestException` with a better error message instead? Seems like a user error, so shouldn't respond with 500?

anwar6953 replied:
yea, the caller can catch it and respond with 400:
![image](https://cloud.githubusercontent.com/assets/2440977/16347487/d137f524-3a01-11e6-97f6-67bc14b9f2a1.png)

Edit: perhaps you're suggesting we should directly throw the BadRequestException from here?

bdmogal replied:
yes, why rely on a try...catch in client code? Unless you think that's beyond what `parseBody`'s contract should be. But if we're going to have this exact same try block everywhere, it perhaps make sense to throw a `BadRequestException` from here itself. `parseBody` will never be used outside of HTTP Handler code right?

anwar6953 replied:
There's places calling this, unrelated to my code, so filed this JIRA for this task:
https://issues.cask.co/browse/CDAP-6296

My PR doesn't use this method; it simply cleaned up surrounding code.

bdmogal replied:
ok 👍

bdmogal commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/remote/RemoteRuntimeStore.java:
```diff
@@ -0,0 +1,127 @@
+  @Override
+  public void compareAndSetStatus(Id.Program id, String pid, ProgramRunStatus expectedStatus,
+                                  ProgramRunStatus updateStatus) {
+    executeRequest("compareAndSetStatus",
```
nit: can be a single line: everywhere in this file?

bdmogal commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/remote/RemoteRuntimeStore.java:
```diff
@@ -0,0 +1,127 @@
+    doPost("execute/" + methodName, GSON.toJson(createRequest(arguments)));
+  }
+
+  private List<MethodArgument> createRequest(Object... arguments) {
```
`createArguments`, since it only deals with arguments?

bdmogal commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/remote/RemoteRuntimeStore.java:
```diff
@@ -0,0 +1,127 @@
+
+  private List<MethodArgument> createRequest(Object... arguments) {
+    List<MethodArgument> methodArguments = new ArrayList<>();
+    for (Object param : arguments) {
```
nit: either use parameters or arguments but not both?

bdmogal commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/remote/RemoteStoreClient.java:
```diff
@@ -0,0 +1,114 @@
+    this.httpRequestConfig = new HttpRequestConfig(httpClientTimeoutMs, httpClientTimeoutMs);
+  }
+
+  private InetSocketAddress getAppFabricServiceAddress() {
```
this method is only called once, let's remove it and add its contents to `resolve`?

bdmogal commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/remote/RemoteStoreClient.java:
```diff
@@ -0,0 +1,114 @@
+  }
+
+  protected HttpResponse doPost(String resource, String body) {
+    return doRequest(resource, "POST", null, body);
```
use constant for POST?

bdmogal commented on cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/store/RemoteRuntimeStoreTest.java:
```diff
@@ -0,0 +1,148 @@
+    RunRecordMeta runMeta = store.getRun(flowId, pid);
+    Assert.assertEquals(initialRunRecord, runMeta);
+
+
```
nit: extra newline

bdmogal commented on cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/store/RemoteRuntimeStoreTest.java:
```diff
@@ -0,0 +1,148 @@
+
+
+    runtimeStore.setSuspend(flowId, pid);
+    Assert.assertEquals(new RunRecordMeta(initialRunRecord, null, ProgramRunStatus.SUSPENDED),
```
so the pattern in this test is to use `RuntimeStore` to set stuff and `Store` to read it back for verification. Will we ever need the ability to read stuff from the store remotely?

anwar6953 replied:
We don't currently, which is why we separated out Store interface into two different interfaces:
Store and RuntimeStore.
This allows us to keep the remote APIs minimal.

bdmogal commented on cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/store/RemoteRuntimeStoreTest.java:
```diff
@@ -0,0 +1,148 @@
+
+  @Test
+  public void testWorkflowMethods() {
+    Id.Workflow workflowId = Id.Workflow.from(Id.Namespace.DEFAULT, "test_app", "test_workflow");
```
can we use the new classes? Or is there a different PR for it? Don't want to add new deprecated code.

anwar6953 replied:
Will come in separate cleanup PR. It's task #3 in pending items.

Nishith commented on cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/meta/RemoteRuntimeStoreHandler.java:
```diff
@@ -0,0 +1,171 @@
+  public void compareAndSetStatus(HttpRequest request, HttpResponder responder) throws ClassNotFoundException {
+    List<MethodArgument> arguments = parseArguments(request);
+
+    Id.Program program = deserialize(arguments.get(0));
```
Perhaps better to use named constants in place of magic numbers in each method? At least for the common ones.

anwar6953 replied:
used iterator to make the code cleaner

bdmogal commented on cdap-proto/src/main/java/co/cask/cdap/proto/BasicThrowable.java:
```diff
@@ -85,7 +87,41 @@ public String getMessage() {
+    if (!Arrays.equals(stackTraces, that.stackTraces)) {
+      return false;
+    }
+    return !(cause != null ? !cause.equals(that.cause) : that.cause != null);
```
will this be easier:

`return cause != null ? cause.equals(that.cause) : that.cause == null`?

anwar6953 replied:
you're right; that does read more easily.
This is just what intellij auto-generated. Updated.

bdmogal commented:
@anwar6953 mostly looks good. Still don't understand why the `WorkflowNodeThrowable*` classes had to be changed + renamed?

anwar6953 commented:
I changed the RuntimeStore APIs to accept the WorkflowNodeThrowable, instead of a generic Throwable. I did this because all the parameters of RuntimeStore API have to be serializable.

Since the method that accepts it is generic (for all programs), and there's nothing specific to workflow in that class, I renamed it to something more generic.

Nishith commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/remote/RemoteRuntimeStore.java:
```diff
@@ -0,0 +1,127 @@
+  }
+
+  private void executeRequest(String methodName, Object... arguments) {
+    doPost("execute/" + methodName, GSON.toJson(createRequest(arguments)));
```
Just curious, don't we need to handle the failure case?

anwar6953 replied:
It is kind of fatal, so I'm not sure what to do in case of failure.
Perhaps we should retry. @chtyim ?

Nishith commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/remote/RemoteRuntimeStore.java:
```diff
@@ -0,0 +1,118 @@
+    doPost("execute/" + methodName, GSON.toJson(createArguments(arguments)));
+  }
+
+  private List<MethodArgument> createArguments(Object... arguments) {
```
Turn in to private static?

Nishith commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/remote/RemoteStoreClient.java:
```diff
@@ -0,0 +1,111 @@
+import javax.annotation.Nullable;
+
+/**
+ * Implements functionality common to making requests from implementations of remote stores.
```
I have no idea what this means :)

bdmogal replied:
How about "Common HTTP client functionality for remote store implementations"?

bdmogal commented:
LGTM

awholegunch commented on cdap-app-fabric/src/main/java/co/cask/cdap/app/store/RuntimeStore.java:
```diff
@@ -0,0 +1,111 @@
+   * Implementation of this method should guarantee that the operation is atomic or in transaction.
+   *
+   * @param id              id of the program
+   * @param pid             The run id
```
Begin param descriptions in lower-case, and with a single space between the param and the description.

awholegunch commented on cdap-app-fabric/src/main/java/co/cask/cdap/app/store/RuntimeStore.java:
```diff
@@ -0,0 +1,111 @@
+   * @param id         id of the program
+   * @param pid        run id
+   * @param startTime  start timestamp in seconds; if run id is time-based pass the time from the run id
+   * @param twillRunId twill run id
```
Twill (as it is a name)

awholegunch commented on cdap-app-fabric/src/main/java/co/cask/cdap/app/store/RuntimeStore.java:
```diff
@@ -0,0 +1,111 @@
+
+  /**
+   * Logs suspend of a program run.
+   * @param id      id of the program
```
insert blank line

awholegunch commented on cdap-app-fabric/src/main/java/co/cask/cdap/app/store/RuntimeStore.java:
```diff
@@ -0,0 +1,111 @@
+
+  /**
+   * Logs resume of a program run.
+   * @param id      id of the program
```
insert blank line

awholegunch commented on cdap-app-fabric/src/main/java/co/cask/cdap/app/store/RuntimeStore.java:
```diff
@@ -0,0 +1,111 @@
+  /**
+   * Add node state for the given {@link Workflow} run. This method is used to update the
+   * state of the custom actions started by Workflow.
+   * @param workflowRunId the Workflow run
```
insert blank line

anwar6953 commented:
Addressed @awholegunch's doc comments.
Got LGTM from @bdmogal on the code earlier, so squashing.
Will merge after unit tests and integration tests pass.

anwar6953 commented:
Unit tests passed here: http://builds.cask.co/browse/CDAP-DUT4241-4
Integration test passed here: http://builds.cask.co/browse/CDAP-ITM-4
